Bernoulli(1) = +½

Parcly-Taxel opened the pull request:
#### References to other Issues or PRs

See below.

#### Brief description of what is fixed or changed

This PR was split into multiple sets:

- [x] #23973
- [x] #23984
- [x] #23985
- [x] #23989
- [x] #23992
- [x] #23994
- [x] #24010
- [x] #24012
- [x] #24014
- [x] #24024

#### Other comments

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

Parcly-Taxel committed:
combinatorial: Correct definition of Bernoulli/Genocchi numbers

Bernoulli's original definition of his numbers had B(1) = +1/2, yet
SymPy currently has B(1) = -1/2. This commit corrects the discrepancy,
as well as the related Genocchi numbers (G(1) = -1, G(0) = 0).

The reference used for these two functions is Peter Luschny's
"An introduction to the Bernoulli function", available at
https://arxiv.org/abs/2009.06743.

Fixes sympy/sympy#23866.

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -423,18 +423,18 @@ class bernoulli(Function):
 
     They are also commonly defined by their exponential generating
     function, which is `\frac{x}{e^x - 1}`. For odd indices > 1, the
     Bernoulli numbers are zero.
```
The generating function would need to be changed.

Parcly-Taxel committed:
Correct overlooked exponential generating function

oscarbenjamin commented:
We also need to check `evalf`, `lambdify` (with mpmath, scipy, ...). Possibly other things. All usage of bernoulli in the codebase should be reviewed (`git grep bernoulli`).

oscarbenjamin commented:
The test failure for Python 3.11 optional dependencies is unrelated to the changes in this PR.

asmeurer commented:
It's good to see that there aren't any other test failures, although we should confirm that there aren't uses of bernoulli that aren't now invalid. 

I still think it could be useful to provide a flag to get the alternate convention, but that might not be so easy to do. I guess we could also just document that the alternative convention can be obtained with `(-1)**n*bernoulli(n)`.

It also would be good to write a note in the docstring about this, noting the convention that is used and why, and noting that it was changed.

asmeurer commented:
Please write a more descriptive release notes entry. This isn't a "bug", it's a change in convention, and it's a breaking change too so it needs to be highlighted as such (especially since the change is being made without depreciation).

Parcly-Taxel committed:
functions: Fix inconsistency of zeta(-n,a) with Apostol's Intro to ANT

When n is a non-negative integer Apostol's Introduction to Analytic
Number Theory (Theorem 12.13 in the 1995 edition) states that the
Hurwitz zeta function ζ(-n, a) may be reduced to a simpler expression
involving a Bernoulli polynomial:

           -B     (a)
             n + 1
ζ(-n, a) = ───────────
              n + 1

This is valid for any complex a, so for example ζ(-1, -4) = -121/12,
agreeing with mpmath. However, SymPy and (catastrophically) Mathematica
return 119/12 for these arguments. This commit introduces the above
relation to SymPy, simplifying a small piece of code, as well as
correcting two test cases.

With this change, ζ(s, a) has only one branch cut along the negative
real axis for fixed noninteger s instead of infinitely many, as can be
seen by plotting in Mathematica.

Parcly-Taxel commented:
> Please write a more descriptive release notes entry. This isn't a "bug", it's a change in convention, and it's a breaking change too so it needs to be highlighted as such (especially since the change is being made without depreciation).

[Done.](https://twitter.com/Parcly_Taxel/status/1558679399050149888)

Parcly-Taxel committed:
Add docstring for Bernoulli change

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -423,18 +423,27 @@ class bernoulli(Function):
+    .. math :: B_n = -n\zeta(1-n)
+
+    Older versions of SymPy and some other sources define `B_1 = -\frac12`.
+    The resulting sequence can be obtained with ``(-1)**n*bernoulli(n)``.
```
I think it would be good to use Sphinx's `.. versionchanged` at the bottom of the docstring and include a reference to Luschny's Bernoulli manifesto or some other document that discusses the different conventions:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionchanged

Parcly-Taxel committed:
Use a bit more Sphinx syntax for Bernoulli docstring

oscargus commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -457,13 +457,23 @@ class bernoulli(Function):
     * ``bernoulli(n, x)`` gives the nth Bernoulli polynomial in `x`, `B_n(x)`
 
+    .. versionchanged:: 1.11
+        ``bernoulli(1)`` gives `+\frac12` instead of `-\frac12`.
```
```suggestion
        ``bernoulli(1)`` gives `+\frac{1}{2}` instead of `-\frac{1}{2}`.
```
Sure, the current works for this particular case, but I think this is much clearer in general.

Parcly-Taxel replied:
(NB: Stripping all not-strictly-needed braces and such is the way I write LaTeX in my Mathematics Stack Exchange answers.)

asmeurer replied:
For `\frac` the braces help make it a little more human readable, which matters for docstrings since many people will read them there.

oscargus commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -1245,14 +1257,14 @@ class genocchi(Function):
     relation:
 
-    .. math:: \frac{2t}{e^t + 1} = \sum_{n=1}^\infty \frac{G_n t^n}{n!}
+    .. math:: \frac{-2t}{1 + e^{-t}} = \sum_{n=0}^\infty \frac{G_n t^n}{n!}
```
Maybe add a `.. versionchanged` here as well.

Parcly-Taxel committed:
Brace fraction parts

Co-authored-by: Oscar Gustafsson <oscar.gustafsson@gmail.com>

Parcly-Taxel committed:
Add versionchanged for Genocchi numbers

Parcly-Taxel committed:
functions: Refactor Bernoulli function evaluation

In the previous version of bernoulli()'s eval() the check for sym being
None appeared several times in the code, even though it is the most
salient feature differentiating between the Bernoulli numbers and
polynomials. This refactor puts that check near the root of the decision
tree and incidentally also allows suppressing explicitly provided sym=1,
renamed to x=1.

Parcly-Taxel committed:
functions: Refactor zeta function evaluation

Luschny's introduction to the Bernoulli function, when it introduces the
generalised Bernoulli function as the analogue of the Hurwitz zeta
function, notes that the extra parameter "can be any complex number that
is not a nonpositive integer". The implication is that ζ(s,a) is
undefined if a is a nonpositive integer, *except* when s is also a
nonpositive integer, the latter allowing us to apply Apostol's identity.

On these grounds the decision tree for evaluating zeta() has been
refactored, removing the vestiges of two-argument Zeta[] behaviour and
making things like zeta(2,-2) strictly undefined. This new version also
suppresses explicit a=1 arguments like bernoulli().

Parcly-Taxel committed:
Entire Bernoulli function

Parcly-Taxel committed:
Add bernoulli_entire to Sphinx index

Parcly-Taxel committed:
functions: Implement rewriting and expand methods for zeta

Parcly-Taxel committed:
functions: Unify bernoulli_entire() with bernoulli()

Parcly-Taxel committed:
Quash bernoulli_entire()

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -451,21 +466,33 @@ class bernoulli(Function):
+        This choice of value confers several theoretical advantages [5]_,
+        including the extension to complex parameters described above
+        which this function now implements. The previous behavior can
+        be obtained with ``(-1)**n*bernoulli(n)``.
```
It's perhaps worth clarifying that the "previous behaviour" was only defined for nonnegative integer n.

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -451,21 +466,33 @@ class bernoulli(Function):
+    cuts `\frac{2}{3}` of the terms. For Bernoulli polynomials, we use
+    the formula in the definition.
 
     * ``bernoulli(n)`` gives the nth Bernoulli number, `B_n`
+    * ``bernoulli(s)`` gives the Bernoulli function `\operatorname{B}(s)`
     * ``bernoulli(n, x)`` gives the nth Bernoulli polynomial in `x`, `B_n(x)`
+    * ``bernoulli(s, a)`` gives the generalized Bernoulli function
+      `\operatorname{B}(s, a)`
```
It might be good to make explicit that the distinction between `n` and `s` above is that `n` is supposed to represent a nonnegative integer whereas `s` is an arbitrary complex number.

oscarbenjamin replied:
Maybe the text above already makes this clear...

oscarbenjamin commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -554,6 +539,17 @@ def _eval_is_finite(self):
         if arg_is_one is not None:
             return not arg_is_one
 
+    def _eval_expand_func(self, **hints):
+        s = self.args[0]
+        a = self.args[1] if len(self.args) == 2 else S.One
+        if a.is_integer:
+            if a.is_positive:
+                return zeta(s) - harmonic(a-1, s)
+            if a.is_nonpositive and (s.is_integer is False or
+                    s.is_nonpositive is False):
+                return S.NaN
+        return self
```
It doesn't look like all branches of this are tested.

oscarbenjamin commented:
Would it make sense to change the way that this is printed to be more like `B(x, y)`?
```python
In [3]: latex(bernoulli(x, y))
Out[3]: 'B_{x}\\left(y\\right)'

In [4]: print(pretty(bernoulli(x, y)))
B (y)
 x 
```

oscarbenjamin commented:
It seems that evaluation with floats doesn't work:
```python
In [6]: bernoulli(1.2)
---------------------------------------------------------------------------
ValueError

~/current/sympy/sympy.git/sympy/core/evalf.py in evalf_bernoulli(expr, prec, options)
   1050     arg = expr.args[0]
   1051     if not arg.is_Integer:
-> 1052         raise ValueError("Bernoulli number index must be an integer")
   1053     n = int(arg)
   1054     b = mpf_bernoulli(n, prec, rnd)
...
ValueError: Bernoulli number index must be an integer
```

Parcly-Taxel committed:
Fuzz out remaining branches of zeta()'s expansion method

What? harmonic(4.0, -3).evalf() raises an exception? It's 100.0!

Parcly-Taxel commented:
- [ ] Implement [Espinosa–Moll generalisation](https://arblib.org/acb.html?highlight=polygamma#c.acb_polygamma) of the polygamma function to arbitrary complex arguments as its `evalf()`
- [ ] Make `harmonic().evalf()` in the general case go through `zeta()`, not `polygamma()`

Parcly-Taxel committed:
functions: Implement Espinosa-Moll polygamma generalisation

With slightly better checking of positiveness and negativeness.
Incidentally the "successor to mpmath" Arb natively includes the
generalisation:
https://arblib.org/acb.html#c.acb_polygamma

Parcly-Taxel committed:
Remove unused import

Parcly-Taxel committed:
functions: Properly extend harmonic() to complex arguments

The rewrite rule from harmonic() to polygamma() is – as hinted by the
presence of (-1)^m in the expression – only valid for positive integer
m. The generally accepted extension of harmonic() to complex arguments
is

H_{n,m} = ζ(m) - ζ(m, n+1)

as long as m ≠ 1. But 1 is a positive integer, so harmonic().evalf() is
still coded to call the polygamma() rewrite if m = 1.

Parcly-Taxel committed:
functions: Fix evalf() for bernoulli()

The work is done as of this commit.

> What do you get if you multiply six by nine?

"Six by nine. Forty two.
That's it. That's all there is.
I always thought something was fundamentally wrong with the universe"

Parcly-Taxel commented:
> It seems that evaluation with floats doesn't work:

**All done.**

oscarbenjamin commented:
I realise this already doesn't work on master but `bernoulli(x).evalf()` where `x` is a symbol should not raise an error.

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -506,49 +540,65 @@ def _calc_bernoulli(n):
+        return Piecewise((1, Eq(n, 0)), (-n * zeta(1-n, x), True))
+
+    def _eval_evalf(self, prec):
+        from mpmath import mp, workprec
```
This import should be at the top of the file.

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -506,49 +540,65 @@ def _calc_bernoulli(n):
+    def _eval_rewrite_as_zeta(self, n, x=1, **kwargs):
+        from sympy.functions.special.zeta_functions import zeta
+        from sympy.functions.elementary.piecewise import Piecewise
+        from sympy.core.relational import Eq
```
Do all of these imports need to be in the function like this rather than at the top of the file?

Parcly-Taxel replied:
Imports in function methods are common throughout the functions module – I see no need to break this pattern. Besides, pushing the imports up to module scope would cause circular import errors, such as with my `zeta()` and `bernoulli()` in different modules referencing each other.

oscarbenjamin replied:
> Imports in function methods are common throughout the functions module

Ideally they wouldn't be. There is an actual runtime cost that comes from having so many delayed imports.

The pattern comes from trying to avoid circular imports but ends up being used in situations where there wouldn't even be any circular imports. I'm pretty sure that here `Eq` can just be imported at the top for example.

Parcly-Taxel replied:
Still, done. When is the earliest possible SymPy release that can include this PR?

oscarbenjamin replied:
It will be 1.12. Probably in a few months.

Parcly-Taxel committed:
Quash error from bernoulli(x).evalf()

Parcly-Taxel committed:
Push mpmath import to top

Parcly-Taxel committed:
More imports moved up top

Parcly-Taxel commented:
> Would it make sense to change the way that this is printed to be more like `B(x, y)`?

That consideration can be deferred until later – it is a minor issue since both the subscripted and non-subscripted forms are common in the literature.

By the way do I need to do anything regarding the failing polygamma import?

oscarbenjamin commented:
> do I need to do anything regarding the failing polygamma import?

No, I fixed that (#23945)

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel committed:
Bump versionchanged version

Parcly-Taxel committed:
Genocchi function

Once the generalised Bernoulli function is defined, a generalised
Genocchi function makes itself clear. As Luschny (2020) defines it:

G(s,a) = 2^s (B(s,a/2) - B(s,(a+1)/2))

Parcly-Taxel committed:
Tests for Genocchi polynomials and a reference

Parcly-Taxel committed:
Euler function at complex arguments

oscarbenjamin commented:
To be clear the more things you put into one PR the harder it is for anyone to review and merge it. I just took a look through the diff and it's now very difficult to:

- See exactly what has changed in terms of behaviour.
- Understand whether all changes are tested.
- Understand whether new tests are the result of changed behaviour or just improving test coverage.
- Check that all changes are appropriately documented.
- Know whether the release notes above are a complete description of the changes in the PR.

I presume that all of these changes are related somehow because all of these functions are related. If it is possible though to make these changes to each function in separate smaller PRs that build one on top of the other that would be better and we can focus on getting each change merged.

In particular it's a lot easier to review and merge a PR if the changes involved are only of one particular type (e.g. only adding tests or only adding a single feature with tests or only changing/extending one function or only refactoring with no behaviour change etc). When different types of change are mixed together though it gets a lot harder to reason about them because it ends up with a large diff but where every single line needs to be very carefully scrutinised to understand why it is being changed or what it does or does not do.

The number one thing that is hard to review when mixed with anything else is a backwards incompatible change. When reviewing a change that is not backwards compatible you want to be sure that you precisely understand what is being changed and what is not being changed. If that is mixed into a large diff then it can very hard to figure out which changes or tests are related to the incompatible change and which are not.

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel commented:
> I presume that all of these changes are related somehow because all of these functions are related. If it is possible though to make these changes to each function in separate smaller PRs that build one on top of the other that would be better and we can focus on getting each change merged.

The first set of changes in this PR has now been separated for your convenience at #23973.

Parcly-Taxel committed:
Fix bug where harmonic(1,pi).evalf() = 0

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel committed:
Restore zero-indexing for Genocchi tests

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel committed:
Splice out extraneous HEAD

Parcly-Taxel committed:
Remove already-merged Genocchi function changes

oscarbenjamin commented:
I tried this script with current master:
```python
import numpy as np
from sympy import *

x, y = symbols('x, y')

funcs = [bernoulli, zeta, harmonic, genocchi]

special_points = [
    0, S.Half, -S.Half, S(3)/2, -S(3)/2, 1, 2, 3, -1, -2, -3,
    pi, pi/2,
    I, -I, 1+I, 1-I, -1+I, -1-I,
    #oo, -oo, I*oo, -I*oo, zoo,
]
generic_points = (list(np.random.randn(5))
                + list(1j*np.random.randn(5))
                + list(np.random.randn(5)+1j*np.random.randn(5)))

points = special_points
#points = generic_points

bad = 0
total = 0
for f1 in funcs:
    for f2 in funcs:
        total += 1
        expr1 = f1(x, y)
        try:
            expr2 = expr1.rewrite(f2)
        except Exception as e:
            print('rewrite failed:', f1, '->', f2)
            1/0
        fine = True
        for p1 in points:
            for p2 in points:
                try:
                    rep = {x:p1, y:p2}
                    if (expr1.subs(rep).n().n(4, chop=True)
                            != expr2.subs(rep).n().n(4, chop=True)):
                        print()
                        print(f1, f2, rep)
                        print(expr1, ' --> ', expr2)
                        print(expr1.subs(rep), expr2.subs(rep))
                        print(expr1.subs(rep).n(5), expr2.subs(rep).n(5))
                        fine = False
                        break
                    expr1.subs(rep).rewrite(f2)
                except (Exception, RecursionError) as e:
                    print(e)
                    print(f1, f2, rep)
                    fine = False
                    break
        if not fine:
            bad += 1

print(bad, 'bad rewrites out of', total)
```
With `points = generic_points` it runs fine indicating that for most generic real, imaginary or complex arguments the different functions can successfully evaluate numerically and that any rewrite rules are consistent. With `points = special_points` though there are a lot failures mostly of numerical evaluation but also of the rewrites in some cases:
```console
$ python bern.py 

bernoulli bernoulli {x: 1/2, y: 0}

bernoulli bernoulli {x: -1/2, y: 0}

bernoulli bernoulli {x: -3/2, y: 0}

bernoulli bernoulli {x: -1, y: 0}

bernoulli bernoulli {x: -2, y: 0}

bernoulli bernoulli {x: -3, y: 0}
logarithm of a negative number
bernoulli bernoulli {x: I, y: -1}
logarithm of a negative number
bernoulli bernoulli {x: -I, y: -1}
logarithm of a negative number
bernoulli bernoulli {x: 1 + I, y: -1}
logarithm of a negative number
bernoulli bernoulli {x: 1 - I, y: -1}
logarithm of a negative number
bernoulli bernoulli {x: -1 + I, y: -1}
logarithm of a negative number
bernoulli bernoulli {x: -1 - I, y: -1}

bernoulli zeta {x: 1/2, y: 0}

bernoulli zeta {x: -1/2, y: 0}

bernoulli zeta {x: -3/2, y: 0}

bernoulli zeta {x: -1, y: 0}

bernoulli zeta {x: -2, y: 0}

bernoulli zeta {x: -3, y: 0}
logarithm of a negative number
bernoulli zeta {x: I, y: -1}
logarithm of a negative number
bernoulli zeta {x: -I, y: -1}
logarithm of a negative number
bernoulli zeta {x: 1 + I, y: -1}
logarithm of a negative number
bernoulli zeta {x: 1 - I, y: -1}
logarithm of a negative number
bernoulli zeta {x: -1 + I, y: -1}
logarithm of a negative number
bernoulli zeta {x: -1 - I, y: -1}

bernoulli harmonic {x: 1/2, y: 0}

bernoulli harmonic {x: -1/2, y: 0}

bernoulli harmonic {x: -3/2, y: 0}

bernoulli harmonic {x: -1, y: 0}

bernoulli harmonic {x: -2, y: 0}

bernoulli harmonic {x: -3, y: 0}
logarithm of a negative number
bernoulli harmonic {x: I, y: -1}
logarithm of a negative number
bernoulli harmonic {x: -I, y: -1}
logarithm of a negative number
bernoulli harmonic {x: 1 + I, y: -1}
logarithm of a negative number
bernoulli harmonic {x: 1 - I, y: -1}
logarithm of a negative number
bernoulli harmonic {x: -1 + I, y: -1}
logarithm of a negative number
bernoulli harmonic {x: -1 - I, y: -1}

bernoulli genocchi {x: 1/2, y: 0}

bernoulli genocchi {x: -1/2, y: 0}

bernoulli genocchi {x: -3/2, y: 0}

bernoulli genocchi {x: -1, y: 0}

bernoulli genocchi {x: -2, y: 0}

bernoulli genocchi {x: -3, y: 0}
logarithm of a negative number
bernoulli genocchi {x: I, y: -1}
logarithm of a negative number
bernoulli genocchi {x: -I, y: -1}
logarithm of a negative number
bernoulli genocchi {x: 1 + I, y: -1}
logarithm of a negative number
bernoulli genocchi {x: 1 - I, y: -1}
logarithm of a negative number
bernoulli genocchi {x: -1 + I, y: -1}
logarithm of a negative number
bernoulli genocchi {x: -1 - I, y: -1}

zeta bernoulli {x: 1/2, y: 0}

zeta bernoulli {x: 3/2, y: 0}

zeta bernoulli {x: 2, y: 0}

zeta bernoulli {x: 3, y: 0}
zeta: too much cancellation
zeta bernoulli {x: pi, y: 0}
zeta: too much cancellation
zeta bernoulli {x: pi/2, y: 0}
logarithm of a negative number
zeta bernoulli {x: I, y: -1}
logarithm of a negative number
zeta bernoulli {x: -I, y: -1}
logarithm of a negative number
zeta bernoulli {x: 1 + I, y: -1}
logarithm of a negative number
zeta bernoulli {x: 1 - I, y: -1}

zeta zeta {x: 1/2, y: 0}

zeta zeta {x: 3/2, y: 0}
zeta: too much cancellation
zeta zeta {x: pi, y: 0}
zeta: too much cancellation
zeta zeta {x: pi/2, y: 0}
logarithm of a negative number
zeta zeta {x: I, y: -1}
logarithm of a negative number
zeta zeta {x: -I, y: -1}
logarithm of a negative number
zeta zeta {x: 1 + I, y: -1}
logarithm of a negative number
zeta zeta {x: 1 - I, y: -1}

zeta harmonic {x: 1/2, y: 0}

zeta harmonic {x: 3/2, y: 0}
zeta: too much cancellation
zeta harmonic {x: pi, y: 0}
zeta: too much cancellation
zeta harmonic {x: pi/2, y: 0}
logarithm of a negative number
zeta harmonic {x: I, y: -1}
logarithm of a negative number
zeta harmonic {x: -I, y: -1}
logarithm of a negative number
zeta harmonic {x: 1 + I, y: -1}
logarithm of a negative number
zeta harmonic {x: 1 - I, y: -1}

zeta genocchi {x: 1/2, y: 0}

zeta genocchi {x: 3/2, y: 0}
zeta: too much cancellation
zeta genocchi {x: pi, y: 0}
zeta: too much cancellation
zeta genocchi {x: pi/2, y: 0}
logarithm of a negative number
zeta genocchi {x: I, y: -1}
logarithm of a negative number
zeta genocchi {x: -I, y: -1}
logarithm of a negative number
zeta genocchi {x: 1 + I, y: -1}
logarithm of a negative number
zeta genocchi {x: 1 - I, y: -1}

harmonic bernoulli {x: -1, y: 1/2}

harmonic bernoulli {x: -2, y: 1/2}

harmonic bernoulli {x: -3, y: 1/2}

harmonic zeta {x: 1/2, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(1/2) nan
0.61371 nan

harmonic zeta {x: -1/2, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(-1/2) nan
-1.3863 nan

harmonic zeta {x: 3/2, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(3/2) nan
1.2804 nan

harmonic zeta {x: -3/2, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(-3/2) nan
0.61371 nan

harmonic zeta {x: 1, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
1 nan
1.0000 nan

harmonic zeta {x: 2, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
3/2 nan
1.5000 nan

harmonic zeta {x: 3, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
11/6 nan
1.8333 nan

harmonic zeta {x: -1, y: 1/2}

harmonic zeta {x: -2, y: 1/2}

harmonic zeta {x: -3, y: 1/2}

harmonic zeta {x: pi, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(pi) nan
1.8727 nan

harmonic zeta {x: pi/2, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(pi/2) nan
1.3145 nan

harmonic zeta {x: I, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(I) nan
0.67187 + 1.0767*I nan

harmonic zeta {x: -I, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(-I) nan
0.67187 - 1.0767*I nan

harmonic zeta {x: 1 + I, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(1 + I) nan
1.1719 + 0.57667*I nan

harmonic zeta {x: 1 - I, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(1 - I) nan
1.1719 - 0.57667*I nan

harmonic zeta {x: -1 + I, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(-1 + I) nan
0.67187 + 2.0767*I nan

harmonic zeta {x: -1 - I, y: 1}
harmonic(x, y)  -->  zeta(y) - zeta(y, x + 1)
harmonic(-1 - I) nan
0.67187 - 2.0767*I nan

harmonic harmonic {x: -1, y: 1/2}

harmonic harmonic {x: -2, y: 1/2}

harmonic harmonic {x: -3, y: 1/2}

harmonic genocchi {x: -1, y: 1/2}

harmonic genocchi {x: -2, y: 1/2}

harmonic genocchi {x: -3, y: 1/2}

genocchi bernoulli {x: 1/2, y: 0}

genocchi bernoulli {x: -1/2, y: 0}

genocchi bernoulli {x: -3/2, y: 0}

genocchi bernoulli {x: -1, y: 0}

genocchi bernoulli {x: -2, y: 0}

genocchi bernoulli {x: -3, y: 0}
logarithm of a negative number
genocchi bernoulli {x: I, y: -1}
logarithm of a negative number
genocchi bernoulli {x: -I, y: -1}
logarithm of a negative number
genocchi bernoulli {x: 1 + I, y: -1}
logarithm of a negative number
genocchi bernoulli {x: 1 - I, y: -1}
logarithm of a negative number
genocchi bernoulli {x: -1 + I, y: -1}
logarithm of a negative number
genocchi bernoulli {x: -1 - I, y: -1}

genocchi zeta {x: 1/2, y: 0}

genocchi zeta {x: -1/2, y: 0}

genocchi zeta {x: -3/2, y: 0}

genocchi zeta {x: -1, y: 0}

genocchi zeta {x: -2, y: 0}

genocchi zeta {x: -3, y: 0}
logarithm of a negative number
genocchi zeta {x: I, y: -1}
logarithm of a negative number
genocchi zeta {x: -I, y: -1}
logarithm of a negative number
genocchi zeta {x: 1 + I, y: -1}
logarithm of a negative number
genocchi zeta {x: 1 - I, y: -1}
logarithm of a negative number
genocchi zeta {x: -1 + I, y: -1}
logarithm of a negative number
genocchi zeta {x: -1 - I, y: -1}

genocchi harmonic {x: 1/2, y: 0}

genocchi harmonic {x: -1/2, y: 0}

genocchi harmonic {x: -3/2, y: 0}

genocchi harmonic {x: -1, y: 0}

genocchi harmonic {x: -2, y: 0}

genocchi harmonic {x: -3, y: 0}
logarithm of a negative number
genocchi harmonic {x: I, y: -1}
logarithm of a negative number
genocchi harmonic {x: -I, y: -1}
logarithm of a negative number
genocchi harmonic {x: 1 + I, y: -1}
logarithm of a negative number
genocchi harmonic {x: 1 - I, y: -1}
logarithm of a negative number
genocchi harmonic {x: -1 + I, y: -1}
logarithm of a negative number
genocchi harmonic {x: -1 - I, y: -1}

genocchi genocchi {x: 1/2, y: 0}

genocchi genocchi {x: -1/2, y: 0}

genocchi genocchi {x: -3/2, y: 0}

genocchi genocchi {x: -1, y: 0}

genocchi genocchi {x: -2, y: 0}

genocchi genocchi {x: -3, y: 0}
logarithm of a negative number
genocchi genocchi {x: I, y: -1}
logarithm of a negative number
genocchi genocchi {x: -I, y: -1}
logarithm of a negative number
genocchi genocchi {x: 1 + I, y: -1}
logarithm of a negative number
genocchi genocchi {x: 1 - I, y: -1}
logarithm of a negative number
genocchi genocchi {x: -1 + I, y: -1}
logarithm of a negative number
genocchi genocchi {x: -1 - I, y: -1}
16 bad rewrites out of 16
```

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

oscarbenjamin commented:
It seems that this hangs:
```python
euler(-1, -1+I).evalf()
```
Varying the arguments by small floating point numbers it seems as if the function should be continuous at this point and given approximately by
```python
In [3]: euler(-1+1e-10, -1+I).evalf()
Out[3]: -0.460778044231282 + 0.27202905474966⋅ⅈ
```

Parcly-Taxel commented:
> It seems that this hangs:

@oscarbenjamin It's a bug with mpmath, not SymPy. Ultimately my `euler._eval_evalf()` code calls `mpmath.polygamma()`, and the very specific arguments you tried are transformed into another set of arguments which hang `mpmath.polygamma()` at certain precisions (but not the default 53-bit one).

The following code hangs in _pure mpmath_, with or without `gmpy2` installed:
```python
from mpmath import mp
mp.prec = 42
mp.polygamma(0, -0.5+0.5j)
```

Parcly-Taxel committed:
Merge branch 'master' into fix-bernoulli-bug

Parcly-Taxel committed:
The End

Parcly-Taxel commented:
Closing and tagging @PeterLuschny, to whom I have sent [this email](https://www.furaffinity.net/journal/10323784). His dream has been realised.

oscarbenjamin commented:
Noting your comment there about Sage. I'm not sure that Sage would use SymPy directly for something like `bernoulli(1)`. Probably it would need to be changed there separately.

oscarbenjamin commented:
Looks like Sage has a number of options but not SymPy:
https://github.com/sagemath/sage/blob/08202bc1ba7caea46327908db8e3715d1adf6f9a/src/sage/arith/misc.py#L276

Sage's own implementation uses `bernoulli(1) = -1/2`:
https://github.com/sagemath/sage/blob/08202bc1ba7caea46327908db8e3715d1adf6f9a/src/sage/rings/bernmm/bern_rat.cpp#L252-L257

Parcly-Taxel commented:
> Sage's own implementation uses `bernoulli(1) = -1/2`: https://github.com/sagemath/sage/blob/08202bc1ba7caea46327908db8e3715d1adf6f9a/src/sage/rings/bernmm/bern_rat.cpp#L252-L257

I dug through the code and found that _by default_ Sage uses FLINT for smaller arguments. Therefore I have opened https://github.com/wbhart/flint2/pull/1179.
